core/src/main/java/org/apache/iceberg/avro/AvroWithPartnerByStructureVisitor.java

```diff
@@ -161,4 +186,17 @@ public T map(P sMap, Schema map, T value) {
   public T primitive(P type, Schema primitive) {
     return null;
   }
+
+  // ---------------------------------- Helpers ---------------------------------------------
+
+  private Deque<String> fieldNames = Lists.newLinkedList();
+  private Deque<Schema> parentSchemas = Lists.newLinkedList();
```

Comment:
Instead of updating all visitors, why not add extra callbacks like the visitors for Iceberg schemas? I think that supporting `beforeField` and `afterField` would be a better way to handle this than passing the parent and name around. The implementation to get a field's ID from its parent and name seems a bit awkward compared to adding an ID stack in one visitor.

Reply:
I think I did think of that pattern, and from my notes the main reason I didn't do it is that it won't be as clean as the existing pattern of `before/afterField` in other visitors, as different data structure has different way of retrieving field id information. For example, for fields that are part of the struct, the field id is stored in its own `Schema.Field` so that we can directly pass `field` to `before/afterField` within the for loop when looping through fields; but for map value, the id is stored in the map's schema instead of its own schema, so that `beforeMapValue` should be passed with the parent schema. My thought was that to require different visitors to implement `before/after` with all these different parameters, and to duplicate the various ID retrieval logic among data types in different visitor implementations could be messy to reason about, that keeping them here might actually be cleaner.

Reply:
What about creating a fake `Schema.Field` to pass to the before/after method instead? Another alternative is to pass the field information to the method, like this:

```java
  public beforeField(int fieldId, String name, Schema type);
  public afterField(int fieldId, String name, Schema type);
  public beforeListElement(int elementId, Schema elementType);
  public beforeMapKey(int keyId, Schema keyType);
  public beforeMapValue(int valueId, Schema valueType);
```

I think some variation on this would be better. We want to avoid keeping additional state in all visitors.